iroh-dns-server/src/http/rate_limiting.rs

```diff
@@ -1,21 +1,60 @@
 use std::time::Duration;
 
 use governor::{clock::QuantaInstant, middleware::NoOpMiddleware};
+use serde::{Deserialize, Serialize};
 use tower_governor::{
-    governor::GovernorConfigBuilder, key_extractor::PeerIpKeyExtractor, GovernorLayer,
+    governor::GovernorConfigBuilder,
+    key_extractor::{PeerIpKeyExtractor, SmartIpKeyExtractor},
+    GovernorLayer,
 };
 
+/// Config for http rate limit.
+#[derive(Debug, Deserialize, Serialize, Clone, Default)]
+#[serde(rename_all = "lowercase")]
+pub enum RateLimitConfig {
+    /// Disable rate limit for http server.
+    Disabled,
+    /// Enable rate limit for http server based on the connection peer IP address.
+    /// https://docs.rs/tower_governor/latest/tower_governor/key_extractor/struct.PeerIpKeyExtractor.html
+    #[default]
+    Simple,
+    /// Enable rate limit for http server based on a smart logic for extracting the connection original IP address, useful for reverse proxies.
+    /// https://docs.rs/tower_governor/latest/tower_governor/key_extractor/struct.SmartIpKeyExtractor.html
+    Smart,
+}
+
 /// Create the default rate-limiting layer.
 ///
 /// This spawns a background thread to clean up the rate limiting cache.
-pub fn create() -> GovernorLayer<'static, PeerIpKeyExtractor, NoOpMiddleware<QuantaInstant>> {
+pub fn create(
+    rate_limit_config: RateLimitConfig,
+) -> Option<GovernorLayer<'static, PeerIpKeyExtractor, NoOpMiddleware<QuantaInstant>>> {
+    let use_smart_extractor = match rate_limit_config {
+        RateLimitConfig::Disabled => {
+            tracing::info!("Rate limiting disabled");
+            return None;
+        }
+        // By default apply rate limit
+        RateLimitConfig::Simple => false,
+        RateLimitConfig::Smart => true,
+    };
+
+    tracing::info!("Rate limiting enabled");
+
     // Configure rate limiting:
     // * allow bursts with up to five requests per IP address
     // * replenish one element every two seconds
-    let governor_conf = GovernorConfigBuilder::default()
-        // .use_headers()
-        .per_second(4)
-        .burst_size(2)
+    let mut governor_conf_builder = GovernorConfigBuilder::default();
+    // governor_conf_builder.use_headers()
+    governor_conf_builder.per_second(4);
+    governor_conf_builder.burst_size(2);
+
+    if use_smart_extractor {
+        tracing::info!("Rate limiting using smart extractor");
```

Comment:
could we change these two info lines for just one that includes the kind of rate limiting?
as in:
```
tracing::info!("Rate limiting enabled ({rate_limit_config})");
```